### ISSUE TITLE
Binding objects before operations

### ISSUE THREAD

**Issue by zgoethel12** [team-jgloom/jgloom#51]
In our implementation, one GL command is wrapped by each object function.  GL command which take the argument `target` operate on the GL object currently bound to the given target, but we never bind them before calling a GL method.

    GLThing t0;
    GLThing t1;
    t1.doThing(GL_THING, are); // 1: Error, none bound
    t0.bind(GL_THING);
    t0.doThing(GL_THING, arg); // 2: Calls glDoThing on t0
    t1.bind(GL_THING); 
    t0.doThing(GL_THING, are); // 3: Calls glDoThing on t1

In this example, number one should call in`t1` and the rest should call on `t0` regardless of what is bound.

**Comment by ecumene-software** [comment]
I primarily avoided this because of the extra opengl calls, but we can surely implement it into the uploaders, but only once at the start of the call!

**Comment by zgoethel12** [comment]
This breaks the idea of individual containers for each object, since objects can operate on each other. Even if you call on a specific object, it only operates on whatever is already bound.

**Comment by zgoethel12** [comment]
Adding `bind(target);` to the top of GL methods with argument `target` should fix it without an extra bare GL call.

**Comment by zgoethel12** [comment]
Here's another way to look at it: all GL wrappers of a given type are identical and can be used interchangeably excluding bind methods.  This is because the wrapped GL method references LWJGL's bound GL object, and not its own.

**Comment by ecumene-software** [comment]
I've got somewhat of a solution:
`JgloomContainer<O>`
 - `onStateChanged():void`
 - `getInstance():void`
     
Then container specific:

`AbstractGLObjContainer` - extends `JgloomContainer<Obj>`
Calls stateChange in every gl state changing method

1. `GLObjContainer` - extends `AbstractGLContainer`
 - Does nothing in stateChange()
2. `StatefulGLObjContainer` - extends `AbstractGLContainer`
 - Binds on stateChange()

**Comment by ecumene-software** [comment]
For the record, I'm still alive! JGLOOm isn't abandoned, I've just been working on a lot of other very interesting projects. Be prepared for a lot of new and closed issues soon.

**Comment by zgoethel12** [comment]
Same here, I've got a few interesting projects opening up at my organizational page `jibini-media` including a Jave wrapper library for HTML (neat) and a CPU rendering system.  Also, I've redesigned `www.jibini.net` with a mobile version.  Now I have `blog.jibini.net` and `cliff.jibini.net` (home DNS) and plan on opening `chat.jibini.net` which we can use to communicate easily.  I emailed you about communication methods.
